kong/clustering/rpc/concentrator.lua

```diff
@@ -0,0 +1,299 @@
+    local res, err = lconn:wait_for_notification()
+    if not res then
+      if err:sub(-7) == "timeout" then
+        repeat
+          local sql, err = self.sub_unsub:pop(0)
+          if err then
+            return nil, err
+          end
+
+          local _, notifications
+          res, err, _, notifications = lconn:query(sql or "SELECT 1;") -- keepalive
+          if not res then
+            return nil, "query to Postgres failed: " .. err
+          end
+
+          enqueue_notifications(notifications, notifications_queue)
+        until not sql
+
+      else
+        return nil, "wait_for_notification error: " .. err
+      end -- err ends with "timeout"
+
+    else
+      notifications_queue:push(res)
+    end
+  end
+end
+
+
+function _M:start(delay)
+  if not self.worker_id then
+    -- this can not be generated inside `:new()` as ngx.worker.id()
+    -- does not yet exist there and can only be generated inside
+    -- init_worker phase
+    self.worker_id = uuid.generate_v5(kong.node.get_id(),
+                                      tostring(ngx.worker.id()))
+  end
+
+  assert(ngx.timer.at(delay or 0, function(premature)
+    if premature then
+      return
+    end
+
+    local lconn = self.db.connector:connect("write")
+    lconn:settimeout(1000)
+    self.db.connector:store_connection(nil, "write")
+
+    local _, res_or_perr, err = pcall(self._event_loop, self, lconn)
+    -- _event_loop never returns true
+    local delay = math.random(5, 10)
+
+    ngx_log(ngx_ERR, "[rpc] concentrator event loop error: ",
+            res_or_perr or err, ", reconecting in ",
+            math.floor(delay), " seconds")
+
+    local res, err = lconn:disconnect()
+    if not res then
+      ngx_log(ngx_ERR, "[rpc] unable to close postgres connection: ", err)
+    end
+
+    self:start(delay)
+  end))
+end
+
+
+-- enqueue a RPC request to DP node with ID node_id
+function _M:_enqueue_rpc_request(node_id, payload)
+  local sql = string_format(RPC_REQUEST_ENQUEUE_SQL,
+                            self.db.connector:escape_literal(node_id),
+                            self.db.connector:escape_literal(self.worker_id),
+                            5,
+                            self.db.connector:escape_literal(cjson_encode(payload)),
+                            self.db.connector:escape_literal(REQ_CHANNEL_PREFIX .. node_id))
+  return self.db.connector:query(sql)
+end
+
+
+-- enqueue a RPC response from CP worker with ID worker_id
+function _M:_enqueue_rpc_response(worker_id, payload)
+  local sql = string_format("SELECT pg_notify(%s, %s);",
+                            self.db.connector:escape_literal(RESP_CHANNEL_PREFIX .. worker_id),
+                            self.db.connector:escape_literal(cjson_encode(payload)))
+  return self.db.connector:query(sql)
+end
+
+
+-- subscribe to RPC calls for worker with ID node_id
+function _M:_enqueue_subscribe(node_id)
+  return self.sub_unsub:push('LISTEN "' .. REQ_CHANNEL_PREFIX .. node_id .. '";')
+end
```

Comment:
If the current Kong node is `data-plane`, this queued message seems not to be consumed by anyone.

```
For DP,

Kong:init_worker()
...
if kong.clustering then
  if is_http_module then
    if is_data_plane(kong.configuration) then
      timer_at(0, function (premature)
        kong.rpc:connect()
        -> self:_add_socket
         -> self.concentrator:_enqueue_subscribe(socket.node_id)
          -> here
```

My understanding is that `concentrator:sub_unsub` queue can only be consumed by Kong node as control plane, do I miss something?

```
For CP,

if kong.clustering then
  if is_http_module then
    if is_data_plane(kong.configuration) then
    ...
    else
      concentrator.start()
        -> concentrator._event_loop
           -> ....
              concentrator:sub_unsub.pop(0)
```

Reply:
It seems, however, that on the DP node, adding a few unprocessed messages during `Kong:init_worker()` doesn't cause any issues. 0.0